modules/apps/batch-engine/batch-engine-service/src/main/java/com/liferay/batch/engine/internal/util/ItemIndexThreadLocal.java

```diff
@@ -16,32 +16,38 @@
 
 import com.liferay.petra.lang.CentralizedThreadLocal;
 
-import java.util.HashMap;
-import java.util.Map;
+import java.util.LinkedList;
+import java.util.Queue;
 
 /**
  * @author Matija Petanjek
  */
 public class ItemIndexThreadLocal {
 
-	public static int get(Object item) {
-		Map<Object, Integer> itemIndexMap = _itemIndexMap.get();
+	public static void add(int itemIndex) {
+		Queue<Integer> indexQueue = _indexQueue.get();
 
-		return itemIndexMap.get(item);
+		indexQueue.add(itemIndex);
 	}
 
-	public static void put(Object item, int itemIndex) {
-		Map<Object, Integer> itemIndexMap = _itemIndexMap.get();
+	public static void clear() {
+		_indexQueue.remove();
+	}
+
+	public static int get() {
+		Queue<Integer> indexQueue = _indexQueue.get();
 
-		itemIndexMap.put(item, itemIndex);
+		return indexQueue.peek();
 	}
 
-	public static void remove() {
-		_itemIndexMap.remove();
+	public static int remove() {
+		Queue<Integer> indexQueue = _indexQueue.get();
+
+		return indexQueue.remove();
 	}
 
-	private static final ThreadLocal<Map<Object, Integer>> _itemIndexMap =
+	private static final ThreadLocal<Queue<Integer>> _indexQueue =
```

Comment:
I think this is a good solution :+1: , we do use ArrayList to store items so their order should be preserved throughout the import task flow.